### PR TITLE
Alternate IGRINS spectra input and some additions to EchelleSpectrumList

### PR DESCRIPTION
**Pull request by kfkaplan**
Added ability to read in IGRINS spectra pipeline products .spec.fits and .wave.fits files.  Also added band math and debasing to EchelleSpectrumList along with a minor bug fix.

### CODE REVIEW

File: src/muler/igrins.py [hunk 1]
```diff
@@ -66,19 +70,29 @@ def __init__(self, *args, file=None, order=10, cached_hdus=None, **kwargs):
                 raise NameError("Cannot identify file as an IGRINS spectrum")
             grating_order = grating_order_offsets[band] + order
 
-            sn_file = file[:-13] + "sn.fits"
+            sn_file = file[19] + "sn.fits"
```
Comment (gully) [inline src/muler/igrins.py:73]: This change breaks the existing code to read in the Signal-to-noise file.   The existing code strips the last 13 characters (`"spec_A0V.fits"`) and replaces them with `"sn.fits"`.  This new code only takes the 19th character and joins it with "sn.fits".

File: src/muler/echelle.py [hunk 2]
```diff
@@ -524,3 +536,35 @@ def plot(self, **kwargs):
             self[i].plot(ax=ax, **kwargs)
 
         return ax
+
+    def __add__(self, other):
+        """Bandmath addition
+        """
+        spec_out = copy.deepcopy(self)
+        for i in range(len(self)):
+            spec_out[i] = self[i] + other[i]
```
Comment (gully) [inline src/muler/echelle.py:545]: This pattern should work!
In the future we may want to test whether the metadata and uncertainties are automatically passed along.  We can make it explicit by saying `spec1.add(spec2, handle_meta='first_found', propagate_uncertainty=True)` rather than `spec1 + spec2` and so on for multiply, divide and subtract.  

We may also want to add some checking for whether the spectral axes are aligned.  If not, bandmath gets weird, and the specutils default may be unexpected.

File: src/muler/igrins.py [hunk 3]
```diff
@@ -42,15 +42,19 @@ class IGRINSSpectrum(EchelleSpectrum):
...
-            sn.fits file, to reduce file I/O for multiorder access.
-            If provided, must give both HDUs.  Optional, default is None.
+            List of two or three fits HDUs, one for the spec.fits/spec_a0v.fits, one for the
+            sn.fits file, and one optional one for the .wave.fits file
+            to reduce file I/O for multiorder access.
+            If provided, must give both (or three) HDUs.  Optional, default is None.
     """
 
-    def __init__(self, *args, file=None, order=10, cached_hdus=None, **kwargs):
+    def __init__(self, *args, file=None, wavefile=None, order=10, cached_hdus=None, **kwargs):
```
Comment (gully) [inline src/muler/igrins.py:57]: I'm slightly inclined to move `wavefile=None` later in the kwargs list.  Most stellar spectroscopists will never need it, and so pushing it later in the list makes it less like to be confused an expected input rather than an optional one.

File: src/muler/igrins.py [hunk 4]
```diff
@@ -66,19 +70,29 @@ def __init__(self, *args, file=None, order=10, cached_hdus=None, **kwargs):
...
             hdr = hdus[0].header
-            lamb = hdus["WAVELENGTH"].data[order].astype(np.float64) * u.micron
-            flux = hdus["SPEC_DIVIDE_A0V"].data[order].astype(np.float64) * u.ct
+            if ".spec_a0v.fits" in file:
+                lamb = hdus["WAVELENGTH"].data[order].astype(np.float64) * u.micron
+                flux = hdus["SPEC_DIVIDE_A0V"].data[order].astype(np.float64) * u.ct
+            elif "spec.fits" in file and wavefile is not None:
+                lamb = wave_hdus[0].data[order].astype(np.float64) * 1e-3 * u.micron #Note .wave.fits and .wavesol_v1.fts files store their wavelenghts in nm so they need to be converted to microns
+                flux = hdus[0].data[order].astype(np.float64) * u.ct
+            else:
```
Comment (gully) [inline src/muler/igrins.py:94]: I recommend adding a case here:

```python
elif ("spec_a0v.fits" in file) and (wavefile is not None):
   log.warn("You have passed in a wavefile and a spec_a0v format file, which has its own wavelength solution.  Ignoring the wavefile.")

elif ("spec.fits" in file) and (wavefile is None):
    raise Exception("wavefile must be specified when passing in spec.fits files, which do not come with an in-built wavelength solution.")
```

File: src/muler/igrins.py [hunk 5]
```diff
@@ -140,24 +154,31 @@ def __init__(self, *args, **kwargs):
         super().__init__(*args, **kwargs)
 
     @staticmethod
-    def read(file, precache_hdus=True):
+    def read(file, wavefile=None, precache_hdus=True):
```
Comment (gully) [inline src/muler/igrins.py:157]: Same as above--- move the `wavefile=None` to the last kwarg, to de-emphasize this pattern for most stellar astronomers.

File: src/muler/igrins.py [hunk 6]
```diff
@@ -66,19 +70,29 @@ def __init__(self, *args, file=None, order=10, cached_hdus=None, **kwargs):
                 raise NameError("Cannot identify file as an IGRINS spectrum")
             grating_order = grating_order_offsets[band] + order
 
-            sn_file = file[:-13] + "sn.fits"
+            sn_file = file[19] + "sn.fits"
             if cached_hdus is not None:
                 hdus = cached_hdus[0]
                 sn_hdus = cached_hdus[1]
+                if wavefile is not None:
+                    wave_hdus = cached_hdus[2]
```
Comment (gully) [inline src/muler/igrins.py:78]: Nice, good catch handling the pre-cached HDU case.  :)

**Review by gully — CHANGES_REQUESTED** [review]
Looks great! See a few comments inline with the code.  The only serious issue is the handling of the `sn.fits` file, where going from `:-13` to `19` breaks the ability to propagate uncertainty.

File: src/muler/igrins.py (hunk 1 above)
Comment (kfkaplan) [reply src/muler/igrins.py:73]: IGRINS pipeline output file names are 19 characters long plus the suffix (e.g. SDCH_20160131_0122.sn.fits).  It should work unless you are planning on compatiblity with renamed files.

File: src/muler/echelle.py (hunk 2 above)
Comment (kfkaplan) [reply src/muler/echelle.py:545]: I tested basic band math and it seems to work.  Propagation of uncertainty and other things are also one of my concerns for specutils (does it automatically do it if a spectrum object has uncertainty?).  As for the wavelength solutions being misaligned, does specutils do some kind of interpolation to align them?

File: src/muler/igrins.py (hunk 3 above)
Comment (kfkaplan) [reply src/muler/igrins.py:57]: Doesn't matter to me really.  Although my understanding is that the .spec.fits files and .wave.fits files are supposed to be the main outputs of the IGRINS pipeline and that the .spec_a0v.fits files were an attempt to try to incorporate a crude telluric correction and relative flux calibration into the pipeline itself.  Are people really grabbing the .spec_a0v.fits for their science work nowadays?

File: src/muler/igrins.py (hunk 1 above)
Comment (kfkaplan) [reply src/muler/igrins.py]: I switched back to the old method and added an if ladder to handle both cases for spec_A0V.fits and spec.fits.

File: src/muler/igrins.py (hunk 3 above)
Comment (kfkaplan) [reply src/muler/igrins.py]: Moved wavefile to the end.

File: src/muler/igrins.py (hunk 5 above)
Comment (kfkaplan) [reply src/muler/igrins.py]: Ditto. Moved wavefile to the end.

File: src/muler/echelle.py (hunk 2 above)
Comment (gully) [reply src/muler/echelle.py:545]: See this tutorial for the answers to your questions--- it's a bit subtle how specutils handles misaligned spectra. https://muler.readthedocs.io/en/latest/tutorials/Combining_uncertainties_with_specutils.html

File: src/muler/igrins.py (hunk 4 above)
Comment (gully) [reply src/muler/igrins.py:102]: I think the elif on line 98 will never trigger, since the first if on line 92 would always be satisfied first.  Maybe move the line 98 elif to an if at the start.

**Review by gully — CHANGES_REQUESTED** [review]

File: src/muler/igrins.py [hunk 7]
```diff
@@ -66,19 +71,36 @@ def __init__(self, *args, file=None, order=10, cached_hdus=None, **kwargs):
                 raise NameError("Cannot identify file as an IGRINS spectrum")
             grating_order = grating_order_offsets[band] + order
 
-            sn_file = file[:-13] + "sn.fits"
+            if ".spec_a0v.fits" in file:
+                sn_file = file[:-13] + "sn.fits"
+            elif ".spec.fits" in file
```
Comment (gully) [inline src/muler/igrins.py:76]: missing a colon `:` at the end

**Review by gully — CHANGES_REQUESTED** [review]

File: src/muler/igrins.py (hunk 4 above)
Comment (kfkaplan) [reply src/muler/igrins.py:102]: Ah good catch.  Moved warning to first if.

File: src/muler/igrins.py (hunk 7 above)
Comment (kfkaplan) [reply src/muler/igrins.py]: Fixed in two places.

**Review by gully — APPROVED** [review]

**Comment by gully** [comment]
Hi @kfkaplan all the changes look great! Also doubly-so thank you for merging the `main` branch in, which allowed the new GitHub Actions to automatically test the code.  All those tests pass, which is a good sign.  

One of your comments about the difference between `spec_A0V.fits` versus `spec.fits` + wavefile made me wonder about the scientific and precision differences between these approaches.  Let's save tha/t thread for another time, and merge this Pull Request, since it appears to do its job with adequate consideration for edge cases.  

I'd like to expand the test suite to test the features added here.  I'm going to add that as a new Issue, again to-be-solved-another day.  

Again thank you for your contribution!! 🙏
